Add section about fargate agents (#30721)
gravitational/teleport: rfd/0109-cloud-agent-upgrades.md

by ec2 discovery.

### ECS Fargate Agents

> When a user uses the Teleport Discover flow to connect an RDS, a Teleport agent is automatically provisioned using ECS Fargate. We need a

Reviewer: To avoid "user uses"
```suggestion
When Teleport Discover flow is used to connect an RDS, a Teleport agent is automatically provisioned using ECS Fargate. We need a
```

> way to keep these agents updated. One idea is to run a separate goroutine in Teleport that periodically checks for an available update and

Reviewer: `goroutine` seems like a too low level implementation detail, maybe we can just call it `separate periodical check`?

> way to keep these agents updated. One idea is to run a separate goroutine in Teleport that periodically checks for an available update and

Reviewer 2: ```suggestion
way to keep these agents updated. Teleport will run a separate goroutine that periodically checks for an available update and
```

> uses the ECS Fargate API to update Teleport agents.

Reviewer: As I understand Fargate instances can't upgrade themselves, that's why we have this separate problem, while other clients in this RFD are supposed to upgrade themselves? Maybe it's worth mentioning?

Author: I've added some more context. The auto updates functionality depends on systemd or Kubernetes, but these are not available for Fargate agents.

This goroutine does not need to be run on all Teleport clusters. It will only be needed if the `TELEPORT_AUTOMATIC_UPGRADES` environment
variable is enabled. The window when Fargate agents are updated should also coincide with the tenant upgrade window.

> The logic flow might look something like:

Reviewer 2: ```suggestion
The logic flow will look like:
```

> - List all available Fargate agents

Reviewer 2: Can you provide more details on how we will determine whether this is a Fargate agent or not? cc @marcoandredinis

- If no agents are available, do nothing
- If agents are available, query the current cloud-stable agent version

> - Update all Fargate agents that do not match the current cloud-stable agent version

Reviewer 2: Can you provide more specifics on how the agents will be updated? Specific AWS API, etc.

> The goroutine will also need to handle critical updates. Periodically, the goroutine will check the critical version endpoint if a critical

Reviewer: I think this paragraph about critical updates should actually be just included into logic flow description. It probably should be very similar to the `core upgrade model` diagram shown above -
1. We check if critical update is available
2. If yes, list all agents and try upgrade them if they are lower version, regardless of maintenance window 
3. If no we check if we're currently inside maintenance window, if no, do nothing
4. If we're in maintenance window list agents and try to upgrade those which need it.

We also will only need to list agents if we're in a maintenance or there's a critical update.
The diagram also mentions aspect of agent healthiness, I'm not sure how it translates into Fargate or if it's applicable, but probably something to consider.

Author: I've updated the flow to match the core upgrade model. They should be pretty much the same.

> The goroutine will also need to handle critical updates. Periodically, the goroutine will check the critical version endpoint if a critical

Reviewer 2: Can we define "periodically"? Once an hour, once every 10 minutes, something else?

Author: The core update model for regular agents loops every 30 minutes. Is it fine if we use the same interval, or do you think we should tune it a bit for Fargate agents?

update is available. If a critical update is available, all Fargate agents will be updated following the same workflow described above. Once

> the critical update is complete, the critical version will need to be stored in memory so that the goroutine does not need to keep listing

Reviewer: Can you add more details about "critical version will need to be stored in memory"? It's a bit unclear.

Author: After thinking about it some more it's probably not necessary. I've removed this detail

Fargate agents to check if they need to be updated.

> To carry out the actual update of a Fargate agent, a new instance will be created with the desired version, then the old instance will be

Reviewer: Will it be possible to do a graceful shutdown of old instance to lower downtime?

Author: I think we should be able to attempt a graceful shutdown. I've added a section about using ECS WaitUntilTasksStopped API to wait for the agent to shutdown before being deleted.

> To carry out the actual update of a Fargate agent, a new instance will be created with the desired version, then the old instance will be

Reviewer 2: Can you clarify what happens when creating a new service fails?

Author: There is a WaitUntilTasksRunning API that we may be able to use to handle new agent failure. We can either retry or just fail and wait for the next update window.

by ec2 discovery.

> ### ECS Fargate Agents

Reviewer 3: I think we should rename the section to refer the integration subkind: AWS OIDC; and the associated action: Deploy Service.
So, maybe something like `AWS OIDC - Deploy Service Agents (Amazon ECS with Fargate)`

When Teleport Discover flow is used to connect an RDS, a Teleport agent is automatically provisioned using ECS Fargate. The automatic update
functionality for regular agents depends on systemd or Kubernetes. Fargate agents run in containers, so that means the existing automatic
update functionality is not compatible with these agents. We need a separate method of keeping these Fargate agents updated.

> To keep Farage agents updated, Teleport will run a separate task that checks every 30 minutes for an available update and uses the ECS

Reviewer 3: Where is that task going to run? I think we should run it on the Proxy

Author: What would be the benefits of running this task on proxy vs auth?

Reviewer 3: We can run this on Auth Service :+1:
My goal was to be explicit about it here, and I usually default to Proxy service

> Fargate API to update Teleport agents. This task does not need to be run on all Teleport clusters. It will only be needed if the

Reviewer 3: Which API are we going to use?

We might need to call the following APIs:
`ecs:DescribeClusters`: the cluster name is deterministic and has some tags that help us ensure it was created by Teleport using the DeployService.
`ecs:DescribeServices`: the service name is deterministic and has some tags that help us ensure it was created by Teleport using the DeployService.

We then need to extract information from the Service, in order for us to recreate it
However, we first need to register a new task definition:
`ecs:DescribeTaskDefinition`: we should get the latest definition used by the service
`ecs:RegisterTaskDefinition`: with the info above, we can create a new task definition that has the new container image version (everything else must stay the same).

Then, we should be ready to re-deploy the service.
We can call `ecs:UpdateService` with the new Task Definition, but I'm not sure it tries graceful shutdown of the container.
If not, then we need `ecs:DeleteService`, `ecs:WaitUntilServicesInactive` and then `ecs:CreateService`

Author: Thanks for giving some more details on the ECS APIs. I'll add them to the RFD.

`TELEPORT_AUTOMATIC_UPGRADES` environment variable is enabled. The window when Fargate agents are updated should also coincide with the
tenant upgrade window.

> The logic flow will look like the same as the core update model:

Reviewer 3: The issue is that if we deploy a bad version, we won't get any heartbeats.

I think we should rely entirely on ECS APIs, except for getting the region which we need to call any AWS API.
How can we get the region? Should we iterate over Teleport (RDS) DBs and create a set of AWS regions, and then call DescribeCluster for each region?

> The logic flow will look like the same as the core update model:

Author: > There's still no section that explains how we will determine that there are agents running in Fargate.

Marco mentioned that `ecs:DescribeClusters` and `ecs:DescribeServices` provides tags that may be used to help identify if the cluster or service was created by Teleport using the DeploymentService. Just added the additional info in the rfd.

> Also, I'm wondering about failure scenarios.

I was thinking we might be able to use `ecs:WaitUntilTasksRunning` and `ecs:WaitUntilTasksStopped` to handle new agent failure and graceful shutdown. I haven't tested if it would work yet, but I was thinking we'd be able to use these APIs to wait for a new agent to be healthy before deleting the existing agent.

> In addition to just checking agent heartbeats, should we also look at Fargate tasks and reconcile them against the cluster state?

I don't fully understand this question. Is this regarding how to check if an agent is in a healthy state?

> The logic flow will look like the same as the core update model:

Reviewer 3: We were discussing the problem of getting all the scoped agents for a given cluster
Let's use the APIs you describe above
For getting the region, let's iterate over all Databases, filter for RDS ones and extract the `AWS.Region`.
For each unique Region, we will call the APIs (DescribeCluster/DescribeServices)

cc @r0mant

- If there is a critical update, attempt update

> - If an agent is unhealthy, attempt update

Reviewer 3: How do we know the agent is unhealthy?

Author: Hmm, do you think we'd be able to use the service/task status to help determine if an agent is healthy/unhealthy?

Reviewer 3: We do not, we must define the HealthCheck for the TaskDefinition.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-healthcheck.html

Then, we must have a command we can run to detect whether Teleport is correctly running (there's no other binary besides `teleport` within the container - we are using the distroless container).

but otherwise, the agent will just be updated during the next maintenance window.

The logic flow will look like the same as the core update model:

- If there is a critical update, attempt update
- If an agent is unhealthy, attempt update
- If within maintenance window, attempt update

[DescribeClusters](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.DescribeClusters) and
[DescribeServices](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.DescribeServices) provide information about the clusters and
services. The cluster and service names are deterministic and the information returned from the API also include tags which can help
identify if the cluster was created by Teleport using the DeployService.

[DescribeTaskDefinition](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.DescribeTaskDefinition) and
[RegisterTaskDefinition](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.RegisterTaskDefinition) can be used to pull the latest
definition and recreate a new task with an updated container image version. Everything else must stay the same.

> To carry out the actual update of a Fargate agent, a new instance will be created with the desired version, then the old instance will be

Reviewer 3: I assume you mean ECS Service when you say "instance", is that it?

shutdown. The workflow will require the [CreateService](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.CreateService), and
[DeleteService](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.DeleteService) ECS API calls.

> We may be able to utilize the [WaitUntilTasksRunning](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.WaitUntilTasksRunning) and

Author: Thanks for pointing that out, I've updated the links to use the v2 sdk
